File: sdks/go/dcp/v2/delegation.go
// Package v2 — DCP-09 Delegation & Representation (Go port).

package v2

import (
	"fmt"
	"time"
)

// DelegationMandateParams bundles inputs for CreateDelegationMandate.
type DelegationMandateParams struct {
	MandateID      string
	SessionNonce   string
	HumanID        string
	AgentID        string
	AuthorityScope []map[string]interface{}
	ValidFrom      string // ISO-8601
	ValidUntil     string // ISO-8601
	Revocable      bool
}

// CreateDelegationMandate produces a mandate signed by the human principal.
// The signature is placed in `human_composite_sig` (not `composite_sig`) to
// distinguish the human-principal witness on the artefact.
func CreateDelegationMandate(
	registry *AlgorithmRegistry,
	humanClassicalKey CompositeKeyInfo,
	humanPQKey CompositeKeyInfo,
	p DelegationMandateParams,
) (map[string]interface{}, error) {
	payload := map[string]interface{}{
		"dcp_version":      "2.0",
		"mandate_id":       p.MandateID,
		"session_nonce":    p.SessionNonce,
		"human_id":         p.HumanID,
		"agent_id":         p.AgentID,
		"authority_scope":  p.AuthorityScope,
		"valid_from":       p.ValidFrom,
		"valid_until":      p.ValidUntil,
		"revocable":        p.Revocable,
		"timestamp":        utcNowISO(),
	}
	canonical, err := CanonicalizeV2(payload)
	if err != nil {
		return nil, fmt.Errorf("canonicalize: %w", err)
	}
	sig, err := CompositeSign(registry, CtxDelegation, []byte(canonical), humanClassicalKey, humanPQKey)
	if err != nil {
		return nil, err
	}
	payload["human_composite_sig"] = sig
	return payload, nil
}

// MandateValidity is the outcome of VerifyMandateValidity.
type MandateValidity struct {
	Valid  bool
	Reason string
}

// VerifyMandateValidity checks mandate expiry and revocation.
func VerifyMandateValidity(mandate map[string]interface{}, revokedMandateIDs map[string]bool) MandateValidity {
	id, _ := mandate["mandate_id"].(string)
	if revokedMandateIDs[id] {
		return MandateValidity{Valid: false, Reason: "Mandate has been revoked"}
	}
	validFrom, _ := mandate["valid_from"].(string)
	validUntil, _ := mandate["valid_until"].(string)
	now := time.Now().UTC()
	vf, err := time.Parse(time.RFC3339, validFrom)
	if err == nil && vf.After(now) {
		return MandateValidity{Valid: false, Reason: "Mandate is not yet valid"}
	}
	vu, err := time.Parse(time.RFC3339, validUntil)
	if err == nil && vu.Before(now) {
		return MandateValidity{Valid: false, Reason: "Mandate has expired"}
	}
	return MandateValidity{Valid: true}
}

// RevocationOutcome is the result of RevokeDelegation.
type RevocationOutcome struct {
	Revoked bool
	Reason  string
}

// RevokeDelegation marks a mandate as revoked, mutating `revokedMandateIDs`.
// Non-revocable mandates return {Revoked: false, Reason: "Mandate is not revocable"}.
func RevokeDelegation(mandate map[string]interface{}, revokedMandateIDs map[string]bool) RevocationOutcome {
	revocable, _ := mandate["revocable"].(bool)
	if !revocable {
		return RevocationOutcome{Revoked: false, Reason: "Mandate is not revocable"}
	}
	if id, ok := mandate["mandate_id"].(string); ok && id != "" {
		revokedMandateIDs[id] = true
	}
	return RevocationOutcome{Revoked: true}
}

// InteractionParams bundles inputs for GenerateInteractionRecord.
type InteractionParams struct {
	InteractionID          string
	SessionNonce           string
	AgentID                string
	CounterpartyAgentID    string
	PublicLayer            map[string]string
	PrivateLayerHash       string
	MandateID              string
}

// GenerateInteractionRecord signs a dual-layer interaction record.
func GenerateInteractionRecord(
	registry *AlgorithmRegistry,
	classicalKey CompositeKeyInfo,
	pqKey CompositeKeyInfo,
	p InteractionParams,
) (map[string]interface{}, error) {
	payload := map[string]interface{}{
		"dcp_version":              "2.0",
		"interaction_id":           p.InteractionID,
		"session_nonce":            p.SessionNonce,
		"agent_id":                 p.AgentID,
		"counterparty_agent_id":    p.CounterpartyAgentID,
		"public_layer":             p.PublicLayer,
		"private_layer_hash":       p.PrivateLayerHash,
		"mandate_id":               p.MandateID,
		"timestamp":                utcNowISO(),
	}
	canonical, err := CanonicalizeV2(payload)
	if err != nil {
		return nil, fmt.Errorf("canonicalize: %w", err)
	}
	sig, err := CompositeSign(registry, CtxDelegation, []byte(canonical), classicalKey, pqKey)
	if err != nil {
		return nil, err
	}
	payload["composite_sig"] = sig
	return payload, nil
}

File: sdks/rust/src/v2/delegation.rs
//! DCP-09 v2.0 Delegation & Representation — Rust port.

use std::collections::HashSet;

use serde_json::{json, Value};

use crate::v2::canonicalize::canonicalize_v2;
use crate::v2::composite_ops::{composite_sign, CompositeKeyInfo};
use crate::v2::crypto_provider::{CryptoError, CryptoProvider};
use crate::v2::domain_separation::CTX_DELEGATION;
use crate::v2::lifecycle::utc_now_iso_pub as utc_now_iso;

pub struct DelegationMandateParams<'a> {
    pub mandate_id: &'a str,
    pub session_nonce: &'a str,
    pub human_id: &'a str,
    pub agent_id: &'a str,
    pub authority_scope: Vec<Value>,
    pub valid_from: &'a str,
    pub valid_until: &'a str,
    pub revocable: bool,
}

/// Create a delegation mandate signed by the human principal.
/// The output carries the signature in the `human_composite_sig` field, not
/// `composite_sig`, to distinguish the human-principal witness on the mandate.
pub fn create_delegation_mandate(
    human_classical_provider: &dyn CryptoProvider,
    human_pq_provider: &dyn CryptoProvider,
    human_classical_key: &CompositeKeyInfo,
    human_pq_key: &CompositeKeyInfo,
    params: DelegationMandateParams<'_>,
) -> Result<Value, CryptoError> {
    let payload = json!({
        "dcp_version": "2.0",
        "mandate_id": params.mandate_id,
        "session_nonce": params.session_nonce,
        "human_id": params.human_id,
        "agent_id": params.agent_id,
        "authority_scope": params.authority_scope,
        "valid_from": params.valid_from,
        "valid_until": params.valid_until,
        "revocable": params.revocable,
        "timestamp": utc_now_iso(),
    });
    let canonical = canonicalize_v2(&payload).map_err(CryptoError::SignatureError)?;
    let composite = composite_sign(
        human_classical_provider,
        human_pq_provider,
        CTX_DELEGATION,
        canonical.as_bytes(),
        human_classical_key,
        human_pq_key,
    )?;
    let mut out = payload;
    if let Some(obj) = out.as_object_mut() {
        obj.insert(
            "human_composite_sig".into(),
            serde_json::to_value(&composite).map_err(|e| CryptoError::SignatureError(e.to_string()))?,
        );
    }
    Ok(out)
}

#[derive(Debug, Clone)]
pub struct MandateValidity {
    pub valid: bool,
    pub reason: Option<String>,
}

/// Check whether a mandate is currently valid (not expired, not revoked).
pub fn verify_mandate_validity(mandate: &Value, revoked_mandate_ids: &HashSet<String>) -> MandateValidity {
    let mandate_id = mandate
        .get("mandate_id")
        .and_then(Value::as_str)
        .unwrap_or("");
    if !mandate_id.is_empty() && revoked_mandate_ids.contains(mandate_id) {
        return MandateValidity {
            valid: false,
            reason: Some("Mandate has been revoked".into()),
        };
    }
    let valid_from = mandate.get("valid_from").and_then(Value::as_str).unwrap_or("");
    let valid_until = mandate.get("valid_until").and_then(Value::as_str).unwrap_or("");
    let now = utc_now_iso();
    // Lexicographic comparison works for ISO-8601 UTC strings.
    if valid_from > now.as_str() {
        return MandateValidity {
            valid: false,
            reason: Some("Mandate is not yet valid".into()),
        };
    }
    if valid_until < now.as_str() {
        return MandateValidity {
            valid: false,
            reason: Some("Mandate has expired".into()),
        };
    }
    MandateValidity {
        valid: true,
        reason: None,
    }
}

#[derive(Debug, Clone)]
pub struct RevocationOutcome {
    pub revoked: bool,
    pub reason: Option<String>,
}

/// Mark a mandate as revoked. Mutates `revoked_mandate_ids`. Non-revocable
/// mandates are returned as `revoked: false` with a reason string.
pub fn revoke_delegation(mandate: &Value, revoked_mandate_ids: &mut HashSet<String>) -> RevocationOutcome {
    if !mandate.get("revocable").and_then(Value::as_bool).unwrap_or(false) {
        return RevocationOutcome {
            revoked: false,
            reason: Some("Mandate is not revocable".into()),
        };
    }
    if let Some(id) = mandate.get("mandate_id").and_then(Value::as_str) {
        revoked_mandate_ids.insert(id.to_string());
    }
    RevocationOutcome {
        revoked: true,
        reason: None,
    }
}

pub struct InteractionParams<'a> {
    pub interaction_id: &'a str,
    pub session_nonce: &'a str,
    pub agent_id: &'a str,
    pub counterparty_agent_id: &'a str,
    pub public_layer: Value,
    pub private_layer_hash: &'a str,
    pub mandate_id: &'a str,
}

pub fn generate_interaction_record(
    classical_provider: &dyn CryptoProvider,
    pq_provider: &dyn CryptoProvider,
    classical_key: &CompositeKeyInfo,
    pq_key: &CompositeKeyInfo,
    params: InteractionParams<'_>,
) -> Result<Value, CryptoError> {
    let payload = json!({
        "dcp_version": "2.0",
        "interaction_id": params.interaction_id,
        "session_nonce": params.session_nonce,
        "agent_id": params.agent_id,
        "counterparty_agent_id": params.counterparty_agent_id,
        "public_layer": params.public_layer,
        "private_layer_hash": params.private_layer_hash,
        "mandate_id": params.mandate_id,
        "timestamp": utc_now_iso(),
    });
    let canonical = canonicalize_v2(&payload).map_err(CryptoError::SignatureError)?;
    let composite = composite_sign(
        classical_provider,
        pq_provider,
        CTX_DELEGATION,
        canonical.as_bytes(),
        classical_key,
        pq_key,
    )?;
    let mut out = payload;
    if let Some(obj) = out.as_object_mut() {
        obj.insert(
            "composite_sig".into(),
            serde_json::to_value(&composite).map_err(|e| CryptoError::SignatureError(e.to_string()))?,
        );
    }
    Ok(out)
}
